src/NHibernate/Context/MapBasedSessionContext.cs

```diff
@@ -20,30 +23,35 @@ protected override ISession Session
 		{
 			get
 			{
-				IDictionary map = GetMap();
-				if (map == null)
-				{
-					return null;
-				}
-				else
-				{
-					return map[_factory] as ISession;
-				}
+				return GetConcreteMap()?.GetOrAdd(_factory, default(ISession));
```

Comment:
More correct as of what this getter has to do, you are right.
A bit more verbose:

```C#
ISession value = null;
GetConcreteMap()?.TryGetValue(_factory, out value);
// We want null if no value was there, no need to explicitly handle false outcome of TryGetValue.
return value;
```

Or:

```C#
var map = GetConcreteMap();
if (map == null)
	return null;
map.TryGetValue(_factory, out var value);
// We want null if no value was there, no need to explicitly handle false outcome of TryGetValue.
return value;
```